File: src/Core/rotor.py
'''
    EnigmaSimulator - A software implementation of the Engima Machine.
    Copyright (C) 2015-2020 Engima Simulator Development Team

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
'''
from rotor_contact import NO_OF_ROTOR_CONTACTS, RotorContact


# ***********************************************************************
# Class representing an Enigma rotor / wheel /drum / Walzen (German).
# ***********************************************************************
class Rotor:
    __slots__ = ['notch_locations', 'reverse_rotor_wiring', 'rotor_name',
                 'rotor_position', 'rotor_ring_setting', 'rotor_wiring']


    ##
    # Property getter : Name of the rotor.
    @property
    def name(self):
        return self.rotor_name

    ##
    # Property getter : How the rotor is wired.
    @property
    def wiring(self):
        return self.rotor_wiring

    ##
    # Property getter 'NotchLocations' : Location of the turnover notch(es).
    @property
    def notches(self):
        return self.notch_locations

    ##
    # Property getter : Position of the rotor.
    @property
    def position(self):
        return self.rotor_position

    @position.setter
    def position(self, value):
        # Validate rotor positions.
        if value < 1 or value > NO_OF_ROTOR_CONTACTS:
            raise ValueError("Invalid rotor positions")

        # Set the position.
        self.rotor_position = value

    ##
    # Property getter 'RingSetting' : Ring setting of the rotor.
    @property
    def ring_setting(self):
        return self.rotor_ring_setting

    @ring_setting.setter
    def ring_setting(self, value):
        if value < 1 or value > NO_OF_ROTOR_CONTACTS:
            raise ValueError("Invalid ring positions")

        self.rotor_ring_setting = value


    ## Rotor constructor method, we wire a rotor from right => left.d.
    # @param self The object pointer.
    # @param name Human readable rotor name.
    # @param wiring Wiring setting from right to left.
    # @param notchLocations Location of the turnover notches.
    def __init__(self, name, wiring, notchLocations):
        # Name of the rotor (e.g. Rotor I).
        self.rotor_name = name

        # Location of the turnover notch(es).
        self.notch_locations = notchLocations

        # Current position of the rotor.
        self.rotor_position = 1

        # Ring setting (Ringstellung) for the rotor
        self.rotor_ring_setting = 1

        # Check to make sure wiring is a list.
        if not isinstance(wiring, (dict)):
            raise ValueError("Incompatible rotor wiring diagram")

        if len(wiring) != NO_OF_ROTOR_CONTACTS:
            raise ValueError("Incomplete wiring diagram")

        # define how the rotor is internally wired.
        self.rotor_wiring = wiring

        self.reverse_rotor_wiring = {v:k for k,v in wiring.items()} 


    ##
    # Step the rotor.
    # @param self The object pointer.
    def step(self):
        if self.rotor_position == NO_OF_ROTOR_CONTACTS:
            self.rotor_position = 1
        else:
            self.rotor_position += 1


    ## Get the output (circuit) using the contacts on the right-hand side
    #  contacts (forward) of the rotor.
    #  @param self The object pointer.
    #  @param contact Reference contact to get circuit with.
    #  @return A contact number.
    def get_forward_circuit(self, contact):

        contact_no = contact.value

        # ===========================================================
        # STEP 1 : Determine the starting point for the input contact
        # ===========================================================
        # This needs to be adjusted for the current position of the rotor.  If
        # the key pressed plus position is greater then NO_OF_ROTOR_CONTACTS
        # then we start at the beginning of the alphabet again.
        # Example 1: 'A' is pressed with position of 'A' will return you the
        #            output from 'A', e.g. for Enigma Rotor 1 will return 'E'
        #            for a letter 'A'.
        # Example 2: 'A' is pressed with position of 'B' will return you the
        #            output from 'B' ('A' has been moved on 1 as now rotor is
        #            in position 'B'), e.g. for Enigma Rotor 1 will return 'K'
        #            for a letter 'B'
        input_contact = (contact_no + self.rotor_position)  -1
        if input_contact > NO_OF_ROTOR_CONTACTS:
            input_contact = input_contact - NO_OF_ROTOR_CONTACTS

        output_contact = self.rotor_wiring[input_contact]

        # ===========================================================
        # STEP 2 : Take ring settings into account
        # ===========================================================
        # CURRENTLY NOT IMPLEMENTED

        # Ring settings are not implemented - untested code
        # if self.__ringSetting > 1:
        #     #outputPin = outputPin + (self.__ringSetting -1)
        #
        #     if (outputPin - (self.__ringSetting -1)) <= 0:
        #         outputPin = NumberOfRotorPins - ((self.__ringSetting -1) \
        #          - outputPin)
        #     else:
        #         outputPin -= (self.__ringSetting -1)
        #
        #     #if outputPin > NumberOfRotorPins:
        #     #    outputPin = outputPin - NumberOfRotorPins;

        final_contact = output_contact

        # ===========================================================
        # STEP 3 : Take rotor offset into account
        # ===========================================================
        # When a rotor has stepped, the offset must be taken into account to
        # know what the output is, and where it enters the next rotor.  If
        # the key pressed plus position is greater then NO_OF_ROTOR_CONTACTS
        # then we start at the beginning of the alphabet again.
        # Example 2: 'A' is pressed with position of 'B' will return you the
        #            output from 'B' as  rotor is in position 'B', e.g. Enigma
        #            Rotor 1 will return 'K' for a letter 'B', but the rotor
        #            is in position 'B' (forward 1) so exit is off by 1, the
        #            output needs to account for this so 'J' is returned.
        # Example 2: 'Z' is pressed with position of 'B' will return you the
        #            output from 'A' as rotor is in position 'B' and this then
        #            wraps about ('Z' forward 1 = 'A'), e.g. Enigma Rotor 1
        #            will return 'E' for a letter 'A', but the rotor is in
        #            position 'B' (forward 1) so exit is off by 1, the output
        #            needs to account for this so 'J' is returned.
        if self.rotor_position > 1:
            if (output_contact - (self.rotor_position -1)) <= 0:
                final_contact = NO_OF_ROTOR_CONTACTS \
                    - ((self.rotor_position -1) - output_contact)

            else:
                final_contact -= (self.rotor_position -1)

        print(f'Final contact: {RotorContact(final_contact)}')
        return final_contact


    ##
    # Get the output (circuit) from the return route using the contacts as a
    # reference.
    # @param self The object pointer.
    # @param contact Reference contact to get circuit with.
    # @return The contact number.
    def get_return_circuit(self, contact):

        contact_no = contact.value
        print(f"::get_return_circuit() Contact : {RotorContact(contact).name}")

        # ===========================================================
        # STEP 1 : Determine the starting point for the input contact
        # ===========================================================
        # This needs to be adjusted for the current position of the rotor.  If
        # the key pressed plus position is greater then NO_OF_ROTOR_CONTACTS
        # then we start at the beginning of the alphabet again.
        # Example 1: 'A' is pressed with position of 'A' will return you the
        #            output from 'A', e.g. for Enigma Rotor 1 will return 'E'
        #            for a letter 'A'.
        # Example 2: 'A' is pressed with position of 'B' will return you the
        #            output from 'B' ('A' has been moved on 1 as now rotor is
        #            in position 'B'), e.g. for Enigma Rotor 1 will return 'K'
        #            for a letter 'B'
        input_contact = contact_no + (self.rotor_position -1)
        if input_contact > NO_OF_ROTOR_CONTACTS:
            input_contact = input_contact - NO_OF_ROTOR_CONTACTS

        # Get the outgoing contact number using the reverse wiriting dictionary.
        output_contact = self.reverse_rotor_wiring[input_contact]
        print(f"::get_return_circuit() output_contact : {RotorContact(output_contact).name}")

        # ===========================================================
        # STEP 2 : Take ring settings into account
        # ===========================================================
        # CURRENTLY NOT IMPLEMENTED

        # if self.__ringSetting > 0:
        #     if (outputPin - (self.__ringSetting -1)) <= 0:
        #         outputPin = NumberOfRotorPins - ((self.__ringSetting -1) \
        #          - outputPin)
        #     else:
        #         outputPin -= (self.__ringSetting -1)
        #
        # Ring settings are not implemented - untested code
        # if self.__ringSetting > 1:
        #     op = outputPin
        #     outputPin = outputPin + (self.__ringSetting -1)
        #     print("[DEBUG] Pin after ring offset : '{0}' [{1}] returns '{2}' [{3}]"
        #     .format(RotorContacts[outputPin], outputPin,
        #      RotorContacts[outputPin], outputPin))
        #
        #     print("[DEBUG] OutputPin = pin ({0}) + ring ({1}) -1 == {2}"
        #     .format(op, self.__ringSetting, outputPin))
        #
        #     if outputPin > NumberOfRotorPins:
        #         outputPin = outputPin - NumberOfRotorPins;

        final_contact = output_contact

        # ===========================================================
        # STEP 3 : Take rotor offset into account
        # ===========================================================
        # When a rotor has stepped, the offset must be taken into account to
        # know what the output is, and where it enters the next rotor.  If
        # the key pressed plus position is greater then NO_OF_ROTOR_CONTACTS
        # then we start at the beginning of the alphabet again.
        # Example 2: 'A' is pressed with position of 'B' will return you the
        #            output from 'B' as  rotor is in position 'B', e.g. Enigma
        #            Rotor 1 will return 'K' for a letter 'B', but the rotor
        #            is in position 'B' (forward 1) so exit is off by 1, the
        #            output needs to account for this so 'J' is returned.
        # Example 2: 'Z' is pressed with position of 'B' will return you the
        #            output from 'A' as rotor is in position 'B' and this then
        #            wraps about ('Z' forward 1 = 'A'), e.g. Enigma Rotor 1
        #            will return 'E' for a letter 'A', but the rotor is in
        #            position 'B' (forward 1) so exit is off by 1, the output
        #            needs to account for this so 'J' is returned.
        if self.rotor_position > 1:
            if (output_contact - (self.rotor_position -1)) <= 0:
                final_contact = NO_OF_ROTOR_CONTACTS \
                    - ((self.rotor_position -1) - output_contact)

            else:
                final_contact -= (self.rotor_position -1)

        print(f'Final contact: {RotorContact(final_contact)}')
        return final_contact


    ##
    # Check to see if the rotor will cause the next one to also step.
    # @param self The object pointer.
    # @return True if when this steps it will cause the next to to, otherwise
    # False is returned.
    def will_step_next(self):
        return self.rotor_position in self.notch_locations

File: src/libEnigmaSimulator/Rotor.cpp
/*
    Engima Machine Simulator
    Copyright (C) 2015-2024 Engima Simulator Development Team

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
*/
#include "Rotor.h"
#include "Logging.h"
#include "StringUtils.h"

namespace enigmaSimulator {

    const int MAX_CONTACT_NO = 26;

    Rotor::Rotor (std::string rotor_name,
        RotorWireConfiguration wiring,
        std::vector<RotorContact> notches,
        RotorContact initialPosition)
        : rotor_name_ (rotor_name),
          notches_ (notches),
          ring_position_(kRotorContact_A),
          rotor_position_(initialPosition)
    {
        if (!wiring.HasValidWiring ())
        {
            throw std::runtime_error ("Wiring layout is not valid");
        }

        wiring_ = wiring;
        wiring_default_ = wiring;
    }

    void Rotor::RotorPosition (RotorContact position)
    {
        rotor_position_ = position;
        RecalculateWiring ();
    }

    // Property getter : Position of the ring.
    void Rotor::RingPosition (RotorContact pos)
    {
        ring_position_ = pos;
        RecalculateWiring ();
    }

    void Rotor::Step ()
    {
        rotor_position_ = RotorContact((rotor_position_ + 1) % MAX_CONTACT_NO);
        RecalculateWiring ();
    }

    /*
    Encrypt a character forward.

    Example 1
    'A' is pressed with the rotor in position 1 ('A'), it will returns the
    output from 'A'. Enigma Rotor 1 will return 'E' for letter 'A'.

    Example 2
    'A' is pressed with the rotor in position 2 ('B'), it will return the
    output from 'B' ('A' has been moved on 1 as rotor is in position 'B').
    Enigma Rotor 1 will return 'K' for a letter 'B'
    */
    RotorContact Rotor::EncryptForward (RotorContact contact)
    {
        int rotor_offset = rotor_position_ - kRotorContact_A;
        DebugLog( "Rotor::EncryptForward", "=> [%s] Input Contact '%c' (%d)",
             rotor_name_.c_str(), RotorContactStr[contact], contact);
        DebugLog( "Rotor::EncryptForward",
                  "=> Current Rotor position : '%c' (%d) (offset : %d)",
            RotorContactStr[rotor_position_], rotor_position_, rotor_offset);

        auto contact_position = contact;

        RotorContact output_contact = kRotorContact_end;

        output_contact = wiring_.WiringPathForward (contact_position);
        DebugLog( "Rotor::EncryptForward", "=> Destination : '%c' (%d)",
            RotorContactStr[output_contact], output_contact);

        rotor_offset = -(rotor_position_ - kRotorContact_A);
        DebugLog( "Rotor::EncryptForward",
            "=> Adjusting destination rotor for offset : '%c' (%d) by %d",
        RotorContactStr[output_contact], output_contact, rotor_offset);

        output_contact = OffsetContactPosition (
            output_contact, -(rotor_position_ - kRotorContact_A));
        DebugLog( "Rotor::" + std::string(__func__),
            "=> Outgoing Rotor position = '%c' (%d)",
        RotorContactStr[output_contact], output_contact);

        return output_contact;
    }

    /*
    Encrypt a character in reverse.

    Example 1
    'A' is pressed with the rotor in position 1 ('A'), it will returns the
    output from 'A'. Enigma Rotor 1 will return 'E' for letter 'A'.

    Example 2
    'A' is pressed with the rotor in position 2 ('B'), it will return the
    output from 'B' ('A' has been moved on 1 as rotor is in position 'B').
    Enigma Rotor 1 will return 'K' for a letter 'B'
    */
    RotorContact Rotor::EncryptReverse (RotorContact contact)
    {
        int rotor_offset = rotor_position_ - kRotorContact_A;
        DebugLog( "Rotor::EncryptReverse",
                  "=> Contact '%c' (%d) | rotor : '%s'",
            RotorContactStr[contact], contact, rotor_name_.c_str());
        DebugLog( "Rotor::" + std::string(__func__),
                  "=> Current Rotor position : '%c' (%d) is offset by %d",
            RotorContactStr[rotor_position_], rotor_position_, rotor_offset);

        // Correct the input contact entrypoint for position
        auto contact_position = OffsetContactPosition (
            contact, rotor_position_ - kRotorContact_A);

        RotorContact output_contact = kRotorContact_end;

        // Get the destination, in this case the 'source'.
        output_contact = wiring_.WiringPathReverse (contact_position);
        DebugLog( "Rotor::EncryptReverse",
            "=> Backwards destination contact for '%c' (%d) : '%c' (%d)",
            RotorContactStr[contact_position],
            contact_position,
            RotorContactStr[output_contact],
            output_contact);

        DebugLog( "Rotor::EncryptReverse",
            "=> Adjusting destination rotor for offset : '%c' (%d) by %d",
        RotorContactStr[output_contact], output_contact, -rotor_position_);

        output_contact = OffsetContactPosition (
            output_contact, -(rotor_position_ - kRotorContact_A));
        DebugLog( "Rotor::EncryptReverse",
            "=> Outgoing Rotor position = '%c' (%d)",
        RotorContactStr[output_contact], output_contact);

        return output_contact;
    }

    /*
        Check to see if this rotor will cause the following also step. It
        returns true if it will cause the next to to, otherwise false.
    */
    bool Rotor::WillStepNext ()
    {
        return std::find (notches_.begin (),
                          notches_.end (), rotor_position_)
                          != notches_.end ();
    }

    RotorContact Rotor::OffsetContactPosition (RotorContact contact,
                                               const int offset)
    {
        DebugLog( "Rotor::" + std::string(__func__),
                  "=> Offsetting Rotor '%c' (%d) by %d positions",
               RotorContactStr[contact], contact, offset);

        int new_position = contact + offset;

        if (new_position > kRotorContact_Z)
        {
            DebugLog( "Rotor::" + std::string(__func__),
                      "=> Positive rotor offset");
            new_position = (new_position % MAX_CONTACT_NO);
        }
        else if (new_position < kRotorContact_A)
        {
            DebugLog( "Rotor::" + std::string(__func__),
                "=> Negative rotor offset");
            new_position = MAX_CONTACT_NO + new_position;
        }

        DebugLog( "Rotor::" + std::string(__func__),
                  "=> Offsetted rotor is '%c' (%d)",
               RotorContactStr[new_position], new_position);

        return RotorContact(new_position);
    }

    void Rotor::PrettyPrintWiring()
    {
        TraceLog( kLogLevel_trace,
                  "Source      : %s", wiring_.GetSrcWiringPathStr().c_str());
        TraceLog( kLogLevel_trace,
                  "Destination : %s", wiring_.GetDestWiringPathStr().c_str());
    }

    void Rotor::RecalculateWiring ()
    {
        auto src = wiring_default_.GetSrcWiringPathStr ();
        auto dest = wiring_default_.GetDestWiringPathStr ();

        // Stage 1 : Modify wiring based on ring position.
        if (ring_position_ > kRotorContact_A)
        {
            // 1] Get the offset ring (remove 1 to account for position 'a').
            int ring_offset = static_cast<int>(ring_position_)
                - static_cast<int>(kRotorContact_A);
            // 2] Rotate all destination wiring 'ring_offset' places right.
            RightRotateString (dest, ring_offset);
            // 3] Offset/increment all destination wiring by 'ring_offset'.
            //    E.g. with ring_offset of 2, 'A' would become 'C'.
            OffsetStringValue (dest, ring_offset);
        }

        // Stage 2 : Modify wiring based on rotor position.
        // 1] Get the rotor offset (remove 1 to account for position 'a').
        int rotor_offset = static_cast<int>(rotor_position_)
            - static_cast<int>(kRotorContact_A);
        // 2] Rotate all source wiring 'rotor_position_' places left.
        LeftRotateString(src, rotor_offset);
        // 2] Rotate all destination wiring 'rotor_position_' places left.
        LeftRotateString(dest, rotor_offset);

        wiring_ = RotorWireConfiguration(dest, src);
    }

}   // namespace enigmaSimulator
